### ISSUE TITLE
[phase-3] SOPD umbrella — self-training LoRA OPD (训推一体): 自更新 + 升级

### ISSUE THREAD

**Issue by cklxx** [cklxx/agent-infer#90]
**Umbrella** for the SOPD axis — teacher-free self-training LoRA OPD: the model self-updates its own
rank-r adapter *at rollout time* (训推一体), with periodic merge+re-quant (升级 / 流式自动量化).
OPD-only / distillation, **never GRPO** ([2026-05-18 pivot](https://github.com/cklxx/arle/blob/main/docs/projects/2026-05-18-opd-only-pivot.md)).

**Canonical docs** (truth):
- Plan: [`docs/plans/2026-06-14-self-training-lora-opd-sopd.md`](https://github.com/cklxx/arle/blob/main/docs/plans/2026-06-14-self-training-lora-opd-sopd.md)
- Survey (apples-to-apples axis comparison): [`docs/research/2026-06-14-self-training-lora-options-survey.md`](https://github.com/cklxx/arle/blob/main/docs/research/2026-06-14-self-training-lora-options-survey.md)

## Child issues (by complexity + cohesion)

| # | Scope | Cohesion | Gate / dep |
|---|---|---|---|
| **#91** | Phase-0 KEYSTONE — inline G2 A1-EMA loop (CUDA+Qwen3.5) | train OPD loop; EmaSelfTeacher + wiring | **blocks all**; dep #92 |
| **#92** | [kv] prefix-cache epoch invalidation on live weight update (R1 HIGH) | infer-core/seam correctness primitive, base-neutral, reusable | prerequisite for #91 PASS |
| **#93** | Phase-0.5 best-of-N rejection (A2) + GSM8K verifier | OPD verifier + best-of-N wrapper | gated on #91 |
| **#98** | Axis-A5 rubric-graded self-distill (ROPD machinery, Path-A; never GRPO) | rubric Rubricator + blind Verifier; reuses GKD update + `xgrammar-sys` | gated on #93; bridges to #97 |
| **#94** | Phase-1 Metal OPD-driver port (aipc) | Metal backend port | gated on #91 |
| **#95** | Phase-2 quantized-base + fp-adapter serve path (C2) | infer serving kernel path | gated on #91 |
| **#96** | Phase-3 升级 / 流式自动量化 — weight-requant primitive | requant cadence | gated on #95 |
| **#97** | Phase-4 skills data engine + live-serve self-update (C1/C5) | L2 训推一体; splits 4a/4b when reached | gated on whole chain |

```
#92 prefix-cache ──▶ #91 KEYSTONE ──┬──▶ #93 best-of-N ──▶ #98 rubric A5 ──▶ #97 skills + live-serve
                                     ├──▶ #94 Metal port                          ▲
                                     └──▶ #95 quant serve ──▶ #96 升级 ───────────┘
```

**#98 = the open-ended bridge.** #93's exact-match verifier only reaches checkable skills (math/code/tool-schema); #98 swaps in an auto-induced **rubric** Verifier so the self-distill signal extends to **non-verifiable / agentic / writing** skills — which is what #97 actually needs. Adopts the ROPD rubric machinery in **rejection/best-of-N distillation** form (never the published GRPO optimizer); reward-hack is sharper than exact-match, mitigated by blind scoring + EMA-as-judge + the #93 held-out tripwire.

## Decided first cut (ckl 2026-06-14)

- **Timing = G2** (整句后更新 / per-sequence inline) — update fires after each rollout *sequence*; not per-token (G3), not offline batch.
- **Device/scope = CUDA + Qwen3.5-0.8B, 自更新-only** (A1 EMA self-teacher). Prove the loop first (#91).
- **Composition** = A1 EMA + B1 LoRA q/v + C1 fp hot-swap + D1 single-process + G2 inline + H1 base-neutral.

## Why LoRA-only makes it tractable

base frozen+quantized ⇒ backward flows **only through the rank-r adapter** ⇒ (1) AIPC-trainable,
(2) teacher-free (student + self-teacher share the frozen base, ~10 MB adapter per role — kills the
teacher+student co-residency OOM), (3) trivial rollback (base never mutates), (4) quant only on the base, only on 升级.

## Architecture verdict — Phase 0 is genuinely simple (code-grounded)

5 architecture cruxes; **4 of 5 do not bite at L1 (Phase 0)** — proven from source:

| Crux | Bites at | Phase-0 status | Evidence |
|---|---|---|---|
| C1 graph-pointer staleness | **#97 serve only** | **dormant** | L1 rollout is eager `forward_token_logits` (`infer_student.rs:108`), no captured-graph replay |
| C2 merge-in-place only | **#95 quant-serve only** | **non-issue** | train store applies adapters **unmerged** (`lora.rs:186-191`) |
| C3 train/infer = separate CUDA memory | every sync | **accepted** | adapter-only D2H/H2D ~10 MB, off the L1 perf critical path |
| C4 rollout KV not tape-recorded | always | **designed-around** | pass-3 recomputes with tape on |
| C5 batch weight-version skew | **#97 multi-request serve only** | **non-issue** | L1 loop serial; control closure drains before re-merge (`lib.rs:404-441`) |

**Phase-0 net-new = zero new kernels** + one elementwise EMA op (#91) + one prefix-cache epoch hook (#92) + one needle-gate driver (#91). The dangerous cruxes (C1, C5) are **structurally deferred to #97**, not waved away.

## Risks (two HIGH, both silent — split by phase)

- **R1 prefix-cache stale-epoch KV** (HIGH) → **#92** (Phase-0 prerequisite).
- **R6 graph-pointer staleness on live re-merge** (HIGH) → **#97** (Phase-4 only; L1 rollout is eager).
- R2 partial rollback poisons EMA (snapshot {student, AdamW, EMA} together) · R3 EMA divergence (α≈0.999 + needle-gate revert) · R4 BF16 skew (existing canary) · R5 two-pass loses graph speed (accept) · R8 reward-hack (#93 tripwire).

## Priority placement

**Phase-3 axis** (master strategy v2 §3 Phase 3 #3 OPD-reclaims-GPU + #5 AIPC). Under D4 (strict serial)
it does **not** pre-empt the active Phase-1 batched-lane keystone (#60). The Phase-0 premise-test (#91) is a
**cheap CUDA probe runnable on the pod off the critical path**; everything after it is gated on a PASS.
Related: #64 (OPD GPU experiments resume + multi-seed gate), #71 (AIPC route / the #94 Metal port), umbrella #55.

**Comment by cklxx** [comment]
Status update (2026-06-21): umbrella tracking healthy — children #91-#98 filed; recent rubric-OPD progress landing (`8c3e0298`, `cf36e1ee`). Phase 0 code complete (#91), rubric infra in (#98/#93). Consolidating #64's OPD-resume progress under this family.

**Comment by cklxx** [comment]
**Cleanup 2026-07-31:** children #91/#92/#94 closed; **#93/#95/#96/#97/#98 still open** (verified). All 5 are Path-A distillation gates awaiting a ≥2-dim held-out multi-seed PASS bar — no code blocker, they need capability-curve runs. Close this umbrella when the five resolve.

**Comment by cklxx** [comment]
**Dep-graph re-anchor 2026-08-14.** Body's gating diagram is stale on three nodes: #92 (prefix-cache epoch invalidation) closed 2026-07-13, #91 (Phase-0 keystone) closed 2026-06-14, #94 (Metal port) closed 2026-07-13. Remaining open children: #93, #95, #96, #97, #98 — #93 and #95 are no longer blocked. Target re-anchor (ckl 2026-06-16, superseding the Qwen3.5-0.8B framing in the body): real target = DSv4-Flash teacher via InferTeacher → Qwen3.6-35B-A3B student, converging with QAT into quantization-aware distillation; 35B SOPD (EMA self-teacher at 35B) stays a target. Current OPD capacity is on the agent-OPD / rubric-OPD lane (27B-FP8, cc-harness execution-reward RFT) and 256K train enablement (#188/#189); SOPD phases queue behind those.
